### DISCUSSION TITLE
Multiple devices for bleKeyboard based of scene

### DISCUSSION THREAD

**Discussion by baerrs** [OMOTE-Community/OMOTE-Firmware#71] [Q&A]
I have a FireTV and an AppleTV.  They both can use the bluKeyboard just fine.  
The problem is I can only have one of the two devices connected, and currently there isn't a way to have more than one.
I want to give them both unique IDs, I only 1 active at a time, I have both a FireTV and an AppleTV, and they both work perfectly fine with my bluKeyboard. However, the issue is that I can only connect one device at a time. Unfortunately, there doesn't seem to be a way to disconnect from one device and connect to another without having to go through the pairing process all over again. 

My idea is to assign unique IDs to each device so that I can easily switch between them depending on the selected scene. However, after doing some research, I couldn't find any option to disconnect from one device and connect to another without having to repeat the entire pairing process. based on the selected scene.

**Comment by KlausMu** [comment]
Correct, currently this is not possible because of a limitation of the library used.
I see two ways to achieve this, both need some research.

1. disconnect and repair, with the need to pick the correct device (each device will try to reconnect immediately)
- https://github.com/T-vK/ESP32-BLE-Keyboard/issues/147

2. have more than one device connected at the same time

- https://github.com/T-vK/ESP32-BLE-Keyboard/issues/236
- https://github.com/T-vK/ESP32-BLE-Keyboard/issues/95
- https://github.com/T-vK/ESP32-BLE-Keyboard/issues/76
- https://github.com/T-vK/ESP32-BLE-Keyboard/issues/66
- https://github.com/ServAlex/ESP32-BLE-Combo

The library itself does not seem to be maintained very well. Last update is 2 years ago, 174 open issues. And it seems the maintainer does not even know himself how to do it. But I think there is no better library available.

Are you able to do some research and to provide the results here?

**Comment by baerrs** [comment]
Yes I'm looking into this.

**Comment by KlausMu** [comment]
What I found out so far:

nr 2. will not work. There is also a limitation in NimBLE. Only one NimBLEDevice is allowed (maybe due to a limitation of the hardware), therefore only one NimBLEServer is allowed, and a NimBLEServer can not send different data to different clients.

nr 1. (disconnect and repair) could work. NimBLE supports a whitelist for connections (see example "NimBLE_Server_Whitelist.ino" from NimBLE). You need to know the MAC addresses of the devices (FireTV, AppleTV) and set them in a whitelist filter, so that only clients on the whitelist can connect. You could get the bluetooth MAC address of your devices on first connect in the log.

So for this to work, you would need to find out the MAC addresses of your FireTV and AppleTV. These have to be put into the code.
Based on the scene, the BLE keyboard would initiate a new advertisement with the correct whitelist set so that only the AppleTV or FireTV would be accepted.

The BLE keyboard library needed to be changed, there is currently nothing in there that supports something like this.

Do you think the whole approach is feasibible? To put the MAC addresses of the devices into the code to the correct scenes?

**Comment by KlausMu** [comment]
https://github.com/h2zero/NimBLE-Arduino/issues/651

**Reply by baerrs** [reply]
I've researched this issue, and it does not seem possible to do this currently. I'm currently looking into implementing an OMOTE BleKeyboard satellite via MQTT. I've gotten it to work, but the latency is up to 10 seconds, which makes the implementation unusable.  I've opened an issue with BleKeyboard, hopefully I'll come up with a solution.  

https://github.com/T-vK/ESP32-BLE-Keyboard/issues/296

**Reply by KlausMu** [reply]
WiFi and BLE at the same time is only possible when using NimBLE. This was already discussed here: https://github.com/CoretechR/OMOTE/discussions/48
So my guess is that you haven't activated it yet. This is where the "delay" is coming from. Probably it is a WiFi reconnect. You could see it when activating `-D CORE_DEBUG_LEVEL=ARDUHAL_LOG_LEVEL_DEBUG`

You could also take this project as a starting point. Should be possible to turn it into a BLE satellite with minimal effort:
https://github.com/KlausMu/esp32-mqtt-keyboard

**Comment by KlausMu** [comment]
Finally I got it running with support of [h2zero ](https://github.com/h2zero).
If you are interested on how to do it, please read [this](https://github.com/h2zero/NimBLE-Arduino/issues/651#issuecomment-2192368168)

Currently it is only a proof of concept. For having it in OMOTE, the following needs to be done
- extension of the "ESP32 BLE keyboard" library. Different kind of advertisement needs to be supported, and disconnect of a connected client.
- in OMOTE, for each BLE media player (or other BLE device) you need to define the MAC address. MAC address can be seen in the log while first pairing
- add code in OMOTE that switches the BLE connection to the right media player (if needed) before sending a BLE keystroke

Technically the problem is solved. The rest is only a finger exercise.

I'm wondering if someone is still waiting for the support of several BLE devices?
